File: web/src/api/endpoints/log.ts
import type { InfiniteData } from '@tanstack/react-query';
import { useInfiniteQuery, useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient, API_BASE_URL } from '../client';
import { logger } from '@/lib/logger';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

/**
 * 尝试状态
 */
export type AttemptStatus = 'success' | 'failed' | 'circuit_break' | 'skipped';

/**
 * 单次渠道尝试信息
 */
export interface ChannelAttempt {
    channel_id: number;
    channel_key_id?: number;
    channel_name: string;
    model_name: string;
    attempt_num: number;    // 第几次尝试
    status: AttemptStatus;
    duration: number;       // 耗时(毫秒)
    sticky?: boolean;
    msg?: string;
    // [fork] channel key display info
    channel_key_index?: number;
    channel_key_preview?: string;
    channel_key_remark?: string;
}

/**
 * 日志数据
 */
export interface RelayLog {
    id: number;
    time: number;                // 时间戳
    request_model_name: string;  // 请求模型名称
    channel: number;             // 实际使用的渠道ID
    channel_name: string;        // 渠道名称
    actual_model_name: string;   // 实际使用模型名称
    input_tokens: number;        // 输入Token
    output_tokens: number;       // 输出Token
    ftut: number;                // 首字时间(毫秒)
    use_time: number;            // 总用时(毫秒)
    cost: number;                // 消耗费用
    request_content: string;     // 请求内容
    response_content: string;    // 响应内容
    error: string;               // 错误信息
    attempts?: ChannelAttempt[]; // 所有尝试记录
    total_attempts?: number;     // 总尝试次数
    // [fork] channel key & user apikey info
    api_key_id?: number;
    api_key_name?: string;
    channel_key_preview?: string;
    channel_key_remark?: string;
    channel_key_index?: number;
    // [fork] summary payload marker
    content_omitted?: boolean;
}

export type LogScope = 'admin' | 'apikey';

/**
 * 日志列表查询参数
 */
export interface LogListParams {
    page?: number;
    page_size?: number;
    start_time?: number;
    end_time?: number;
}

const logDetailQueryKey = (scope: LogScope, id: number) => ['log', 'detail', scope, id] as const;

// [fork] fetch full log detail on demand
export function useLogDetail(options: { id: number; scope?: LogScope; enabled?: boolean }) {
    const { id, scope = 'admin', enabled = true } = options;
    const logApiBase = scope === 'apikey' ? '/api/v1/apikey/log' : '/api/v1/log';

    return useQuery({
        queryKey: logDetailQueryKey(scope, id),
        queryFn: async () => {
            return apiClient.get<RelayLog>(`${logApiBase}/detail/${id}`);
        },
        enabled: enabled && id > 0,
        staleTime: Infinity,
    });
}

/**
 * 清空日志 Hook
 * 
 * @example
 * const clearLogs = useClearLogs();
 * 
 * clearLogs.mutate();
 */
export function useClearLogs() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async () => {
            return apiClient.delete<null>('/api/v1/log/clear');
        },
        onSuccess: () => {
            logger.log('日志清空成功');
            queryClient.invalidateQueries({ queryKey: ['logs'] });
        },
        onError: (error) => {
            logger.error('日志清空失败:', error);
        },
    });
}

const logsInfiniteQueryKey = (scope: LogScope, pageSize: number) => ['logs', scope, 'infinite', pageSize] as const;

/**
 * 日志管理 Hook
 * 整合初始加载、SSE 实时推送、滚动加载更多
 * 
 * @example
 * const { logs, isConnected, hasMore, isLoadingMore, loadMore, clear } = useLogs();
 * 
 * // logs 自动包含历史日志和实时日志，按时间倒序
 * logs.forEach(log => console.log(log.request_model_name));
 * 
 * // 滚动到底部时加载更多
 * if (hasMore && !isLoadingMore) loadMore();
 */
export function useLogs(options: { pageSize?: number; scope?: LogScope } = {}) {
    const { pageSize = 20, scope = 'admin' } = options;
    const logApiBase = scope === 'apikey' ? '/api/v1/apikey/log' : '/api/v1/log';

    const [isConnected, setIsConnected] = useState(false);
    const [error, setError] = useState<Error | null>(null);
    const eventSourceRef = useRef<EventSource | null>(null);

    const queryClient = useQueryClient();

    const logsQuery = useInfiniteQuery({
        queryKey: logsInfiniteQueryKey(scope, pageSize),
        initialPageParam: 1,
        queryFn: async ({ pageParam }) => {
            const params = new URLSearchParams();
            params.set('page', String(pageParam));
            params.set('page_size', String(pageSize));
            // [fork] fetch lightweight list payload, detail content is loaded on demand
            params.set('include_content', 'false');
            const result = await apiClient.get<RelayLog[] | null>(`${logApiBase}/list?${params.toString()}`);
            return result ?? [];
        },
        getNextPageParam: (lastPage, allPages) => {
            if (!lastPage || lastPage.length < pageSize) return undefined;
            return allPages.length + 1;
        },
        staleTime: Infinity,
        refetchOnMount: 'always',
    });

    const logs = useMemo(() => {
        const pages = logsQuery.data?.pages ?? [];
        const seen = new Set<number>();
        const merged: RelayLog[] = [];

        for (const page of pages) {
            for (const log of page) {
                if (seen.has(log.id)) continue;
                seen.add(log.id);
                merged.push(log);
            }
        }

        merged.sort((a, b) => b.time - a.time);
        return merged;
    }, [logsQuery.data]);

    const loadMore = useCallback(async () => {
        if (!logsQuery.hasNextPage) return;
        if (logsQuery.isFetchingNextPage) return;

        try {
            await logsQuery.fetchNextPage();
        } catch (e) {
            logger.error('加载更多日志失败:', e);
        }
    }, [logsQuery]);

    useEffect(() => {
        let cancelled = false;

        const connect = async () => {
            try {
                const { token } = await apiClient.get<{ token: string }>(`${logApiBase}/stream-token`);
                if (cancelled) return;

                const eventSource = new EventSource(`${API_BASE_URL}${logApiBase}/stream?token=${token}`);
                eventSourceRef.current = eventSource;

                eventSource.onopen = () => {
                    setIsConnected(true);
                    setError(null);
                };

                eventSource.onmessage = (event) => {
                    try {
                        const log: RelayLog = JSON.parse(event.data);
                        queryClient.setQueryData(
                            logsInfiniteQueryKey(scope, pageSize),
                            (old: InfiniteData<RelayLog[], number> | undefined) => {
                                if (!old) {
                                    return { pages: [[log]], pageParams: [1] };
                                }

                                const exists = old.pages.some((p) => p?.some((x) => x.id === log.id));
                                if (exists) return old;

                                const firstPage = old.pages[0] ?? [];
                                return { ...old, pages: [[log, ...firstPage], ...old.pages.slice(1)] };
                            }
                        );
                    } catch (e) {
                        logger.error('解析日志数据失败:', e);
                    }
                };

                eventSource.onerror = () => {
                    setIsConnected(false);
                    setError(new Error('SSE 连接断开'));
                    eventSource.close();
                    eventSourceRef.current = null;
                };
            } catch (e) {
                if (cancelled) return;
                setError(e instanceof Error ? e : new Error('获取 stream token 失败'));
                logger.error('获取 stream token 失败:', e);
            }
        };

        connect();

        return () => {
            cancelled = true;
            eventSourceRef.current?.close();
            eventSourceRef.current = null;
            setIsConnected(false);
        };
    }, [logApiBase, pageSize, queryClient, scope]);

    const clear = useCallback(() => {
        queryClient.removeQueries({ queryKey: logsInfiniteQueryKey(scope, pageSize) });
    }, [pageSize, queryClient, scope]);

    return {
        logs,
        isConnected,
        error,
        hasMore: !!logsQuery.hasNextPage,
        isLoading: logsQuery.isLoading,
        isLoadingMore: logsQuery.isFetchingNextPage,
        loadMore,
        clear,
    };
}

File: internal/model/log.go
package model

// AttemptStatus 尝试状态
type AttemptStatus string

const (
	AttemptSuccess      AttemptStatus = "success"       // 转发成功
	AttemptFailed       AttemptStatus = "failed"        // 转发失败
	AttemptCircuitBreak AttemptStatus = "circuit_break" // 熔断跳过
	AttemptSkipped      AttemptStatus = "skipped"       // 其他原因跳过（禁用、无Key、类型不兼容等）
)

// ChannelAttempt 记录单次渠道尝试的决策和结果
type ChannelAttempt struct {
	ChannelID    int           `json:"channel_id"`
	ChannelKeyID int           `json:"channel_key_id,omitempty"`
	ChannelName  string        `json:"channel_name"`
	ModelName    string        `json:"model_name"`
	AttemptNum   int           `json:"attempt_num"`
	Status       AttemptStatus `json:"status"`
	Duration     int           `json:"duration"`
	Sticky       bool          `json:"sticky,omitempty"`
	Msg          string        `json:"msg,omitempty"`
	// [fork] channel key display info
	ChannelKeyIndex   int    `json:"channel_key_index,omitempty"`
	ChannelKeyPreview string `json:"channel_key_preview,omitempty"`
	ChannelKeyRemark  string `json:"channel_key_remark,omitempty"`
}

type RelayLog struct {
	ID                      int64   `json:"id" gorm:"primaryKey;autoIncrement:false"`    // Snowflake ID
	Time                    int64   `json:"time"`                                        // 时间戳（秒）
	RequestModelName        string  `json:"request_model_name"`                          // 请求模型名称
	ChannelId               int     `json:"channel"`                                     // 实际使用的渠道ID
	ChannelName             string  `json:"channel_name"`                                // 渠道名称
	ActualModelName         string  `json:"actual_model_name"`                           // 实际使用模型名称
	InputTokens             int     `json:"input_tokens"`                                // 输入Token
	OutputTokens            int     `json:"output_tokens"`                               // 输出 Token
	Ftut                    int     `json:"ftut"`                                        // 首字时间(毫秒)
	UseTime                 int     `json:"use_time"`                                    // 总用时(毫秒)
	Cost                    float64 `json:"cost"`                                        // 消耗费用
	RequestContent          string  `json:"request_content"`                             // 请求内容
	OriginalRequestContent  string  `json:"original_request_content"`                    // 原始入站请求内容
	OriginalRequestProtocol string  `json:"original_request_protocol" gorm:"default:''"` // 原始入站协议名
	OutboundRequestContent  string  `json:"outbound_request_content"`                    // 实际出站请求内容
	OutboundRequestProtocol string  `json:"outbound_request_protocol" gorm:"default:''"` // 实际出站协议名
	ResponseContent         string  `json:"response_content"`                            // 响应内容
	// [fork] streaming final preview rendered into a non-stream JSON body for readability
	StreamPreviewContent string           `json:"stream_preview_content" gorm:"default:''"`
	Error                string           `json:"error"`                           // 错误信息
	Attempts             []ChannelAttempt `json:"attempts" gorm:"serializer:json"` // 所有尝试记录
	TotalAttempts        int              `json:"total_attempts"`                  // 总尝试次数
	// [fork] raw upstream response body before protocol conversion
	OriginalResponseContent string `json:"original_response_content"` // 原始上游响应内容
	// [fork] channel key & user apikey info for log display
	APIKeyID          int    `json:"api_key_id" gorm:"default:0;index"`     // user apikey id
	APIKeyName        string `json:"api_key_name" gorm:"default:''"`        // user apikey name
	ChannelKeyPreview string `json:"channel_key_preview" gorm:"default:''"` // masked channel key
	ChannelKeyRemark  string `json:"channel_key_remark" gorm:"default:''"`  // channel key remark
	ChannelKeyIndex   int    `json:"channel_key_index" gorm:"default:0"`    // 1-based index in channel
	// [fork] summary payload marker
	ContentOmitted bool `json:"content_omitted,omitempty" gorm:"-"`
}
